lib/src/builder/mod.rs

```diff
@@ -111,13 +146,42 @@ pub trait BlockBuilderStrategy {
     /// Builds a block from the given input.
     fn build_from(
         chain_spec: &ChainSpec,
-        input: Input<Self::TxEssence>,
-    ) -> Result<(Header, MptNode)> {
-        BlockBuilder::<MemDb, Self::TxEssence>::new(chain_spec, input)
-            .initialize_database::<Self::DbInitStrategy>()?
-            .prepare_header::<Self::HeaderPrepStrategy>()?
-            .execute_transactions::<Self::TxExecStrategy>()?
-            .finalize::<Self::BlockFinalizeStrategy>()
+        input: BlockBuildInput<Self::TxEssence>,
+    ) -> Result<BlockBuildOutput> {
+        // Database initialization failure does not mean the block is faulty
+        // todo: compute only on build error
+        let input_hash = input.state_input.hash();
+        let initialized = BlockBuilder::<MemDb, Self::TxEssence>::new(chain_spec, input, None)
+            .initialize_database::<Self::DbInitStrategy>()?;
+
+        // Header validation errors mean a faulty block
+        let prepared = match initialized.prepare_header::<Self::HeaderPrepStrategy>() {
+            Ok(builder) => builder,
+            Err(_) => {
+                return Ok(BlockBuildOutput::FAILURE {
+                    state_input_hash: input_hash.into(),
+                })
+            }
+        };
+
+        // Transaction execution errors mean a faulty block
```

Comment:
For actual error distinction, we would need to properly define a `ThisError' or something. The execution layer may return errors that are not failures, but other rust errors.

Reply:
I'm not sure we can properly maintain a definitive distinction at that level